Rethink settings screen confirmation

GRGool opened the pull request:
### Removing Save and Cancel Buttons in Settings Screen

**This PR includes the following changes:**

### **1. Refactored Profile and Username Fields into Reusable Composables**
- Introduced `EditableUsernameField` and `EditableProfilePicture` composables to modularize logic and reduce duplication.

---

### **2. Removed Save and Cancel Buttons**
- Logic previously tied to save and cancel buttons has been replaced by context-sensitive confirmation dialogs:
  - **Username Confirmation**: Triggered when the user moves away from the field or presses enter.
  - **Profile Picture Confirmation**: 
    - Displayed after selecting a new picture from the gallery.
    - Displayed when deleting the profile picture.

---

### **3. Added and Adapted Tests**
- Updated existing tests to match the new popup confirmation logic.
- Added tests for the following scenarios:
  - Confirmation dialog appearing after username change and focus loss.
  - Confirmation dialog for profile picture edit and delete actions.
  - Ensured correct updates to the ViewModel and repository when confirmed.
  - Verified that changes are not saved when canceling the confirmation popup.

---

### **4. Focus Handling Improvements**
- Fixed issues related to focus on screen load.
- Ensured that focus-related events only trigger once the user interacts with the screen.

---

### **5. Code Cleanup and Consistency**
- Unified UI behavior for both username and profile picture.
- Formatted the popups to be similar to the other popups of the app.

---

### **Summary of Improvements**
- Enhanced user experience through clear and context-sensitive interactions.
- Simplified code structure with reusable components.
- Addressed focus and interaction issues to make the settings screen behavior more intuitive.

ThomasMPont reviewed (changes requested):
This is only a partial review. I will review your code once the following requests are addressed.

## Missing features

Your PR does not accomplish all of the requirements of your original task.

### Figma

All changes to the UI of a screen must be reflected on Figma.

### Editing usernames

Upon editing my username and clicking outside the username field, I would expect the confirmation popup to show up. I would also expect it to show up when attempting to leave the screen via the top bar or the system bar's back buttons.
In short, there should be no way to avoid having to confirm or cancel changes to the username.
I don't know how difficult this is to implement, but maybe you can have the field intercept any user input while it is selected.

## Bugs

Your PR also contains several bugs and awkward behavior.

### Profile picture

The profile picture editing box should display the user's current profile picture. At the moment, it shows nothing until a new picture is uploaded.

### Deleting nothing

Even when a user has no profile picture. They can still delete their profile picture, asking them for confirmation. I suggest you hide the delete button from users with no profile picture to avoid this.

This is not an issue with usernames.

### Username cancel

When a username change is cancelled, the username field's content is replaced by the user's current username. This conflicts with the current username being a placeholder value of the field when it is empty.
I suggest you remove the placeholder value, and have the field contain the current username when the screen is loaded instead.

### Popup positioning

The confirmation popup currently hides the profile picture and username, making it difficult to tell what the changes are. This should be improved once these elements are moved to the top of the screen. I still suggest you shift the popup down if you can.

GRGool commented:
Hey, Apologies for the blunder. The focus handling is the first thing I implemented but trying to debug my firebase issue I deleted some things (It was most likely an emulator issue with the gallery images). I have still fixed your other concerns : 

Not sure if or how I fixed it but the current user profile picture is displayed on my end.

It is now not possible to escape confirm or cancel.

It is now not possible to delete an empty profile image (not clickable but still shown).

The username is a little bit tricky because of the values returned by the viewmodel. When the screen is loaded it has value "unknown", if you set newname to this value the tests show very weird behaviour. (I would get concatenated strings like "Updated Nameunknown"). I settled for having an independent newname, which is only what you write. The placeholder is linked to the username and the textbox handles its update dynamically, but its not super important if this value takes weird values in initialization as we use it as a visual feedback.

I moved the Popup down, looks a bit weird to me though.

ThomasMPont commented:
@GRGool 
> Hey, Apologies for the blunder.

No worries

> Not sure if or how I fixed it but the current user profile picture is displayed on my end.

You did fix it. Nice.

> It is now not possible to escape confirm or cancel.

I might have conveyed my thoughts wrong. I mainly wanted the popup to show up no matter what. I don't think your change is bad though.

> It is now not possible to delete an empty profile image (not clickable but still shown).

Did you try hiding it? Using a `modifier` like `SPACE_BETWEEN`, you might be able to.

> The username is a little bit tricky because of the values returned by the viewmodel. When the screen is loaded it has value "unknown", if you set newname to this value the tests show very weird behaviour. (I would get concatenated strings like "Updated Nameunknown"). I settled for having an independent newname, which is only what you write. The placeholder is linked to the username and the textbox handles its update dynamically, but its not super important if this value takes weird values in initialization as we use it as a visual feedback.

This feels fine, for now. Maybe you can have another shot at it after the whole "unknown" fiasco is dealt with.

> I moved the Popup down, looks a bit weird to me though.

This will be looked at in [this task](https://github.com/orgs/Signify-epfl/projects/2/views/1?filterQuery=settings&pane=issue&itemId=90615386&issue=Signify-epfl%7Csignify-app%7C419).

ThomasMPont reviewed (changes requested):
You've fixed almost everything, nice work!
There still remains a couple points from my first comment, that I am not satisfied with.

### Clicking off the username field

Your username field behaves as expected when one clicks off on the screen's background. The user can still bypass this confirmation by clicking on any elements, such as the "back" button. Did you try to block all user input? If this is not feasible, could you at least mention what you've tried.

### Hiding the delete button

It feels weird to have this non functional "delete" button there. Did you try deleting it. Again, if you have, could you explain what's stopped you.

GRGool commented:
Not sure I understand. When prompted to confirm or cancel I cant press anything else. This should be handled in the ConfirmationDialog composable where I dont do anything in the onDissmissRequest function. 
If you mean clicking on the backbutton while writing on the textfield then its a deeper issue. 
Here is how I handled the focus. The annexScaffold doesnt take focus so clicking on the screen didnt do anything. I wrapped it in a box that cleared focus to fix it. The elements outside of the annexscaffold such as the switches I can easily call clearfocus on them, but the backbutton im not quite sure how to handle it as its inside the annexscaffold and I dont want to modify that. I didnt quite get what you meant by blocking user input if you could clarify.

Im sure I can find something for the delete button, give me a minute.

ThomasMPont commented:
> Not sure I understand. When prompted to confirm or cancel I cant press anything else. This should be handled in the ConfirmationDialog composable where I dont do anything in the onDissmissRequest function.

Again this is fine.

> If you mean clicking on the backbutton while writing on the textfield then its a deeper issue.

Yes, I was hoping clicking anywhere outside the field would deselect it and ask to save my changes. Right now, you can just leave the screen while typing a new username and lose your change.

GRGool commented:
Its a headache. Tried multiple things but ultimately you either need to catch the navigation back which I dont think can be handled with our current navigationActions, or you need to change the backbutton but we use the annexscaffold one so we cant change it. If you have any idea. 
What I tried but doesn't work/breaks clean structure: textfield has its own invisible layer that wraps everything and catches all clicks, you display this layer with a flag like isEditing, and then you just handle the focus on this layer. I cant really get this to work with how I modularised edit/profile pic. It could work but not clean. 
At the very least the current behaviour respects fail safe rule and just discards the changes.

ThomasMPont commented on app/src/androidTest/java/com/github/se/signify/ui/screens/profile/SettingsScreenTest.kt:
```diff
@@ -95,93 +98,90 @@ class SettingsScreenTest {
+    composeTestRule.onNodeWithText("Cancel").performClick()
+
+    verify(userRepository, never())
+        .updateProfilePictureUrl(Mockito.anyString(), eq(null), anyOrNull(), anyOrNull())
```
## Requested changes

When checking for `never()` you usually want to be as loose as possible. Does this really need to look for `eq(null)`?

GRGool replied:
Fixed.

ThomasMPont commented on app/src/main/java/com/github/se/signify/ui/screens/profile/SettingsScreen.kt:
```diff
@@ -76,165 +84,253 @@ fun SettingsScreen(
+              .clickable(
+                  onClick = { focusManager.clearFocus() },
+                  indication = null,
+                  interactionSource = remember { MutableInteractionSource() })) {
```
I'm assuming there was no better way to have detect taps on the screen's background.

ThomasMPont commented on app/src/main/java/com/github/se/signify/ui/screens/profile/SettingsScreen.kt:
```diff
@@ -76,165 +84,253 @@ fun SettingsScreen(
-                contentDescription = "Edit Username",
-                tint = MaterialTheme.colorScheme.onBackground)
+    ConfirmationDialog(
+        showDeleteDialog,
```
Why is this parameter not named, when the others are?

ThomasMPont commented on app/src/main/java/com/github/se/signify/ui/screens/profile/SettingsScreen.kt:
```diff
@@ -76,165 +84,253 @@ fun SettingsScreen(
+  if (showDialog.value) {
 
-    Spacer(modifier = Modifier.height(64.dp))
+    Dialog(onDismissRequest = {}) {
```
## Requested changes

I think you misunderstood what I asked you to change earlier. I think you should have `onDismissRequest` act like "cancel" does. Right now, it completely blocks the user from leaving, which can be both dangerous and irritating.

GRGool replied:
Agreed, fixed

ThomasMPont commented on app/src/main/java/com/github/se/signify/ui/screens/profile/SettingsScreen.kt:
Your elements are mostly functional, and look good, but their implementation is very hard to read. From a glance at your editable field `@Composable`s, it is very difficult to tell what is what. Perhaps, helper functions, or limiting your use of modifiers would help.

I feel like you would benefit from looking for simpler solutions to your implementation problems. Remember, issues often stem for overly complicated systems.

ThomasMPont commented on app/src/main/java/com/github/se/signify/ui/screens/profile/SettingsScreen.kt:
```diff
@@ -76,165 +84,253 @@ fun SettingsScreen(
+                    2.dp, MaterialTheme.colorScheme.outlineVariant, RoundedCornerShape(8.dp))
+                .clip(RoundedCornerShape(8.dp))
+                .background(MaterialTheme.colorScheme.background)
+                .padding(horizontal = 24.dp, vertical = 16.dp),
```
Does your padding really need to be so specific?

ThomasMPont commented on app/src/main/java/com/github/se/signify/ui/screens/profile/SettingsScreen.kt:
```diff
@@ -76,165 +84,253 @@ fun SettingsScreen(
+                      .padding(vertical = 8.dp)
+                      .background(MaterialTheme.colorScheme.background)
+                      .onFocusChanged { focusState ->
+                        if (!focusState.isFocused && newName != userName && newName.isNotEmpty()) {
```
This could be moved to a helper function.

ThomasMPont commented on app/src/main/java/com/github/se/signify/ui/screens/profile/SettingsScreen.kt:
```diff
@@ -76,165 +84,253 @@ fun SettingsScreen(
+              contentDescription = "Edit Username",
+              tint = MaterialTheme.colorScheme.onBackground)
+
+          Spacer(modifier = Modifier.width(8.dp))
```
You've already made a `Row()`, do you really need `Spacer`s too?

ThomasMPont commented on app/src/main/java/com/github/se/signify/ui/screens/profile/SettingsScreen.kt:
```diff
@@ -76,165 +84,253 @@ fun SettingsScreen(
+    onConfirm: () -> Unit,
+    onDismiss: () -> Unit,
+    title: String,
+    message: String
```
`title` and `message` could have been merged into a data class.

ThomasMPont commented on app/src/main/java/com/github/se/signify/ui/screens/profile/SettingsScreen.kt:
```diff
@@ -76,165 +84,253 @@ fun SettingsScreen(
+          color = MaterialTheme.colorScheme.surface,
+          modifier =
+              Modifier.widthIn(min = 280.dp, max = 400.dp) // Set the dialog width
+                  .padding(top = 150.dp)) {
```
Did you actually specifically need all these modifiers?

ThomasMPont commented on app/src/main/java/com/github/se/signify/ui/screens/profile/SettingsScreen.kt:
```diff
@@ -76,165 +84,253 @@ fun SettingsScreen(
+    title: String,
+    message: String
+) {
+  if (showDialog.value) {
```
This if statement feels like it should be moved out of your `@Composable()`. It feels weird that you call `ConfirmationDialog()`s everywhere, but they aren't actually displayed most of the time.

ThomasMPont reviewed (changes requested):
## Overview

Your changes look great! The settings screen now behaves in a much more conventional way. This is overall a big improvement to the UX.

## Requested changes

Good job on addressing most of my requests from earlier!
I've requested a couple extra changes below. These requests are marked with this sections title.

## Bugs

### Delete button hiding delay

When a user's profile picture is deleted, the delete button doesn't hide instantly. It seems the button needs a round trip to the user repository before hiding itself.
This should be fixed by having the button adapt to local changes rather than waiting for the server to feed it a new profile picture. If you don't think you know how to fix this quickly, I suggest you open a new task that fixes this.

## Process

### Task

You've closely followed your task's specifications, and haven't overstepped by a bit. Great work!

### Commits

Your commits are very useful for following your development process. Your commit messages don't all follow the same convention though.

### Description

Your description is extremely detailed, and has helped a lot with this review.

## Summary

While your changes are functional and seem robust, you often write more code than you need to. This makes your code vulnerable and hard to understand. In the future try to only write the bare minimum lines of code to achieve your goals. Overly complex code generally tends to be more brittle.

GRGool commented on app/src/androidTest/java/com/github/se/signify/ui/screens/profile/SettingsScreenTest.kt:
```diff
@@ -47,6 +48,14 @@ class SettingsScreenTest {
     userViewModel = UserViewModel(userSession, userRepository)
     val picturePath = "file:///path/to/profile/picture.jpg"
 
+    @Suppress("UNCHECKED_CAST")
```
Done, its a GPT thingy, sorry for that.

GRGool commented on app/src/androidTest/java/com/github/se/signify/ui/screens/profile/SettingsScreenTest.kt:
```diff
@@ -95,93 +98,90 @@ class SettingsScreenTest {
-    composeTestRule.onNodeWithText(newName).assertIsDisplayed()
+    composeTestRule.onNodeWithText(newName).performImeAction()
+
+    composeTestRule.onNodeWithText("Confirm Changes").assertIsDisplayed()
```
Agreed and done.

GRGool commented:
Hey, thanks for the review and the advice. As for the changes requested, I have fixed the bug of the delete button by making it depend on the selectedUri, just like the profile pic, it is now instant. I have also reverted the onDissmissRequest to act just like a cancel, I agree its less invasive. I have implemented the other test related changes as well, removing the cast, adding tags, etc. If you could take a look!

GRGool commented on app/src/androidTest/java/com/github/se/signify/ui/screens/profile/SettingsScreenTest.kt:
```diff
@@ -95,93 +98,90 @@ class SettingsScreenTest {
-    // Verify that updateUserName is not called since the input was empty
     verify(userRepository, never())
-        .updateUserName(Mockito.anyString(), Mockito.anyString(), anyOrNull(), anyOrNull())
+        .updateUserName(Mockito.anyString(), eq(newName), anyOrNull(), anyOrNull())
```
Yep, reverted this.

ThomasMPont reviewed (approved):
Looks good to me. Nice work!
